internal/cmd/marketplace/apply/apply_test.go

```diff
@@ -402,18 +406,120 @@ func TestApplyPrintApplyOutcome(t *testing.T) {
+			Done: true,
+			Items: []marketplace.ApplyResponseItem{
+				{
+					ItemID:           "id1",
+					Name:             "some name 1",
+					Done:             true,
+					Inserted:         true,
+					Updated:          false,
+					ValidationErrors: []marketplace.ApplyResponseItemValidationError{},
+				},
+			},
+		}
+		uploadMockResponse := &marketplace.UploadImageResponse{
+			Location: mockImageURLLocation,
+		}
+
+		mockServer := applyIntegrationMockServer(t, http.StatusOK, applyMockResponse, uploadMockResponse)
+		defer mockServer.Close()
+		clientConfig := &client.Config{
+			Transport: http.DefaultTransport,
+		}
+		clientConfig.Host = mockServer.URL
+		client, err := client.APIClientForConfig(clientConfig)
+		require.NoError(t, err)
+
+		found, err := applyItemsFromPaths(
+			context.Background(),
+			client,
+			mockTenantID,
+			mockPaths,
+		)
+		require.NoError(t, err)
+		require.Contains(t, found, "id1")
+		require.Contains(t, found, "some name 1")
+	})
+}
+
+func TestApplyConcatPathIfRelative(t *testing.T) {
+	t.Run("should concat relative paths", func(t *testing.T) {
+		found := concatPathIfRelative("/some/path/to/file.json", "./image.png")
+		require.Equal(t, "/some/path/to/image.png", found)
+	})
+	t.Run("should return an abs path", func(t *testing.T) {
+		found := concatPathIfRelative("/some/path/to/file.json", "/some/absolute/path/image.png")
+		require.Equal(t, "/some/absolute/path/image.png", found)
+	})
+}
+
+func applyRequestHandler(t *testing.T, w http.ResponseWriter, r *http.Request, statusCode int, mockResponse interface{}) {
+	t.Helper()
+	require.Equal(t, mockURI, r.RequestURI)
+	require.Equal(t, http.MethodPost, r.Method)
+
+	w.WriteHeader(statusCode)
+	resBytes, err := json.Marshal(mockResponse)
+	require.NoError(t, err)
+	w.Write(resBytes)
+}
+
+func applyIntegrationMockServer(t *testing.T, statusCode int, applyMockResponse, uploadMockResponse interface{}) *httptest.Server {
 	t.Helper()
 	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		var isReqOk = assert.Equal(t, mockURI, r.RequestURI) && assert.Equal(t, http.MethodPost, r.Method)
-		if !isReqOk {
-			w.WriteHeader(http.StatusNotFound)
-			require.Fail(t, "unsupported call")
-			return
+		switch r.RequestURI {
+		case fmt.Sprintf(applyEndpoint, mockTenantID):
+			foundBodyBytes, err := io.ReadAll(r.Body)
+			require.NoError(t, err)
+			foundBody := make(map[string]interface{})
+			err = json.Unmarshal(foundBodyBytes, &foundBody)
+			require.NoError(t, err)
+			require.Contains(t, foundBody, "resources")
+			resources := foundBody["resources"].([]interface{})
+
+			// verify that the keys have been replaced correctly
+			require.NotContains(t, resources[0].(map[string]interface{}), imageKey)
+			require.Contains(t, resources[0].(map[string]interface{}), imageURLKey)
+			require.Equal(t, mockImageURLLocation, resources[0].(map[string]interface{})[imageURLKey].(string))
```

Comment:
Might it be better to have a function like "expectPlaceholderKeys..." in order to enhance test readability ?

Reply:
Maybe instead of an helper function it can only need a variable for `resources[0].(map[string]interface{})` to avoid repeating it for the three assertion and using a clear name for what is it inside that array that you are testing?